languagetool-core/src/main/java/org/languagetool/rules/patterns/RuleSet.java

```diff
@@ -0,0 +1,149 @@
+/* LanguageTool, a natural language style checker
+ * Copyright (C) 2020 Daniel Naber (http://www.danielnaber.de)
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+package org.languagetool.rules.patterns;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.languagetool.AnalyzedSentence;
+import org.languagetool.rules.Rule;
+
+import java.util.*;
+
+/**
+ * An object holding a set of rules with an optional possibility to fetch only the ones applicable for a given sentence
+ * (via {@link #rulesForSentence}), to avoid even invoking the definitely inapplicable ones.
+ * The filtering is based on cues provided by the rules, e.g. possible token strings or lemmas in pattern rules.
+ *
+ * @since 5.2
+ */
+@ApiStatus.Internal
+public abstract class RuleSet {
```

Comment:
Could you maybe add a unit test for this class? I think it's a good approach to have a test for any new class.